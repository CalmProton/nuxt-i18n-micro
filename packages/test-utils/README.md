# Nuxt I18n Micro Test Utils

This is a utility library designed to facilitate testing for Nuxt.js applications that use the `nuxt-i18n-micro-core` package. It provides helper functions to handle translations, formatting, and locale switching, simplifying the process of testing internationalization (i18n) functionality.

## Features

- **Translation Helper Functions**: Easily retrieve and interpolate translations with the `t` and `tc` functions.
- **Pluralization**: Handle plural forms of translations with the `tc` function.
- **Locale Management**: Set and get the current locale and locale-specific data (e.g., route name, locale list).
- **Number and Date Formatting**: Format numbers and dates according to the current locale.
- **Translation Merging**: Merge new translations into the existing cache with the `mergeTranslations` function.
- **Locale Switching**: Switch between locales dynamically and update route and path accordingly.

## Installation

You can install the package via npm or yarn:

```bash
npm install nuxt-i18n-micro-test-utils
```

or

```bash
yarn add nuxt-i18n-micro-test-utils
```

## Usage

### Importing the helper functions

To use the utility functions, import them as follows:

```typescript
import { i18nUtils } from 'nuxt-i18n-micro-test-utils'
```

### Example Usage

#### Getting a translation

To retrieve a translation, you can use the `t` function:

```typescript
const translatedValue = i18nUtils.t('welcome_message', { name: 'John' })
console.log(translatedValue)  // Output: translated string with the injected name
```

#### Formatting numbers

Use the `tn` function to format numbers based on the current locale:

```typescript
const formattedNumber = i18nUtils.tn(12345.6789)
console.log(formattedNumber)  // Output: formatted number
```

#### Formatting dates

Use the `td` function to format dates:

```typescript
const formattedDate = i18nUtils.td(new Date())
console.log(formattedDate)  // Output: formatted date
```

#### Pluralization

Handle plural translations with `tc`:

```typescript
const pluralValue = i18nUtils.tc('item_count', 3)
console.log(pluralValue)  // Output: appropriate plural form
```

#### Merging Translations

To add new translations, use the `mergeTranslations` function:

```typescript
i18nUtils.mergeTranslations({ welcome_message: 'Hello, {name}!' })
```

### Locale Management

You can manage the current locale with the following functions:

```typescript
i18nUtils.setLocale('en')  // Set current locale to English
const currentLocale = i18nUtils.getLocale()  // Get the current locale
```

### Setting Translations from JSON

Load translations dynamically using the `setTranslationsFromJson` function:

```typescript
const newTranslations = { welcome_message: 'Welcome!' }
i18nUtils.setTranslationsFromJson('en', newTranslations)
```

## License

MIT License. See the [LICENSE](LICENSE) file for details.

## Author

- **Name**: s00d
- **Email**: Virus191288@gmail.com
- **Website**: [https://s00d.github.io/](https://s00d.github.io/)
