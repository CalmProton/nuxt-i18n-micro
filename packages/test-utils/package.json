{
  "name": "nuxt-i18n-micro-test-utils",
  "version": "1.0.1",
  "description": "",
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "vite build",
    "test": "jest",
    "prepublishOnly": "pnpm run build"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "nuxt-i18n-micro-core": "workspace:*"
  },
  "devDependencies": {
    "vite-plugin-dts": "^4.3.0"
  }
}
