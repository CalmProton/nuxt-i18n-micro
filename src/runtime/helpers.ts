import type { Strategies } from '../types'

export const withPrefixStrategy = (strategy: Strategies) => strategy === 'prefix' || strategy === 'prefix_and_default'

export const isNoPrefixStrategy = (strategy: Strategies) => strategy === 'no_prefix'

export const isPrefixStrategy = (strategy: Strategies) => strategy === 'prefix'

export const isPrefixExceptDefaultStrategy = (strategy: Strategies) => strategy === 'prefix_except_default'

export const isPrefixAndDefaultStrategy = (strategy: Strategies) => strategy === 'prefix_and_default'
