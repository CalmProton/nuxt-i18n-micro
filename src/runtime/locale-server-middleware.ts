import type { H3Event } from 'h3'
import type { Locale, ModuleOptionsExtend } from 'nuxt-i18n-micro-types'
import { detectCurrentLocale } from './utils/locale-detector'
import { useRuntimeConfig } from '#imports'

export interface LocaleInfo {
  current: string
  default: string
  fallback: string
  available: string[]
  locale: Locale | null
  isDefault: boolean
  isFallback: boolean
}

export const useLocaleServerMiddleware = (event: H3Event, defaultLocale?: string, currentLocale?: string): LocaleInfo => {
  const config = useRuntimeConfig(event)
  const { locales, defaultLocale: configDefaultLocale, fallbackLocale } = config.public.i18nConfig as ModuleOptionsExtend

  // Определяем текущую локаль по приоритету
  const detectedLocale = currentLocale || detectCurrentLocale(event, { fallbackLocale, defaultLocale: defaultLocale || configDefaultLocale })

  // Находим конфигурацию текущей локали
  const localeConfig = locales?.find(l => l.code === detectedLocale) ?? null

  // Получаем список доступных локалей
  const availableLocales = locales?.map(l => l.code) ?? []

  // Проверяем, является ли текущая локаль дефолтной
  const isDefault = detectedLocale === (defaultLocale || configDefaultLocale || 'en')

  // Проверяем, является ли текущая локаль fallback
  const isFallback = detectedLocale === (fallbackLocale || defaultLocale || configDefaultLocale || 'en')

  return {
    current: detectedLocale,
    default: defaultLocale || configDefaultLocale || 'en',
    fallback: fallbackLocale || defaultLocale || configDefaultLocale || 'en',
    available: availableLocales,
    locale: localeConfig,
    isDefault,
    isFallback
  }
}
