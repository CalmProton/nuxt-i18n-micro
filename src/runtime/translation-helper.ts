import type { Translations } from './plugins/01.plugin'

const generalLocaleCache: Record<string, Translations> = {}
const routeLocaleCache: Record<string, Translations> = {}
const dynamicTranslationsCaches: Record<string, Translations>[] = []

const serverTranslationCache: Record<string, Map<string, Translations | unknown>> = {}
const serverTranslationInit: Record<string, boolean> = {}

function deepClone<T>(value: T): T {
  if (Array.isArray(value)) {
    return JSON.parse(JSON.stringify(value)) as T
  }
  else if (typeof value === 'object' && value !== null) {
    return JSON.parse(JSON.stringify(value)) as T
  }
  return value
}

function findTranslation<T = unknown>(translations: Translations | null, key: string): T | null {
  let value: string | number | boolean | Translations | unknown | null = translations

  if (translations === null) {
    return null
  }

  if (translations[key]) {
    value = translations[key]
  }
  else {
    const parts = key.split('.')
    for (const part of parts) {
      if (value && typeof value === 'object' && part in value) {
        value = (value as Translations)[part]
      }
      else {
        return null
      }
    }
  }

  if (typeof value === 'object' && value !== null) {
    return deepClone(value) as T
  }

  return (value as T) ?? null
}

export function useTranslationHelper() {
  return {
    hasCache(locale: string, page: string) {
      return (serverTranslationCache[`${locale}:${page}`] ?? new Map<string, Translations | unknown>()).size > 0
    },
    getCache(locale: string, routeName: string) {
      return serverTranslationCache[`${locale}:${routeName}`]
    },
    setCache(locale: string, routeName: string, cache: Map<string, Translations | unknown>) {
      const cacheKey = `${locale}:${routeName}`
      serverTranslationCache[`${locale}:${routeName}`] = cache
      serverTranslationInit[`${locale}:index`] = true
      serverTranslationInit[cacheKey] = true
    },
    mergeTranslation(locale: string, routeName: string, newTranslations: Translations, force = false) {
      if (!force && !routeLocaleCache[`${locale}:${routeName}`]) {
        console.error(`marge: route ${locale}:${routeName} not loaded`)
      }
      routeLocaleCache[`${locale}:${routeName}`] = {
        ...routeLocaleCache[`${locale}:${routeName}`],
        ...newTranslations,
      }
    },
    mergeGlobalTranslation(locale: string, newTranslations: Translations, force = false) {
      if (!force && !generalLocaleCache[`${locale}`]) {
        console.error(`marge: route ${locale} not loaded`)
      }
      generalLocaleCache[locale] = {
        ...generalLocaleCache[locale],
        ...newTranslations,
      }
    },
    hasGeneralTranslation(locale: string) {
      return !!generalLocaleCache[locale]
    },
    hasPageTranslation(locale: string, routeName: string) {
      return !!routeLocaleCache[`${locale}:${routeName}`]
    },
    getTranslation: <T = unknown>(locale: string, routeName: string, key: string): T | null => {
      const cacheKey = `${locale}:${routeName}`
      const cached = serverTranslationCache[cacheKey]?.get(key)
      if (cached) {
        return cached as T
      }

      let result: T | null = null

      for (const dynamicCache of dynamicTranslationsCaches) {
        result = findTranslation<T>(dynamicCache[locale] || null, key)
        if (result !== null) break
      }

      if (!result) {
        result = findTranslation<T>(routeLocaleCache[cacheKey] || null, key)
        ?? findTranslation<T>(generalLocaleCache[locale] || null, key)
      }

      if (result) {
        if (!serverTranslationCache[cacheKey]) {
          serverTranslationCache[cacheKey] = new Map<string, Translations>()
        }

        serverTranslationCache[cacheKey].set(key, result)
      }

      return result
    },
    async loadPageTranslations(locale: string, routeName: string, translations: Translations): Promise<void> {
      const cacheKey = `${locale}:${routeName}`
      routeLocaleCache[cacheKey] = { ...translations }
      serverTranslationInit[cacheKey] = true
    },
    async loadTranslations(locale: string, translations: Translations): Promise<void> {
      generalLocaleCache[locale] = { ...translations }
      serverTranslationInit[`${locale}:index`] = true
    },
  }
}
