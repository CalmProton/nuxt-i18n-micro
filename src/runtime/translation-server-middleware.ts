import type { H3Event } from 'h3'
import { getQuery, getCookie } from 'h3'
import { useTranslationHelper } from 'nuxt-i18n-next-core'
import type { Translations } from 'nuxt-i18n-next-core'

async function fetchTranslations(locale: string): Promise<Translations> {
  try {
    const translations = await $fetch(`/_locales/general/${locale}/data.json`)
    return translations as Translations
  }
  catch (error) {
    console.error(`Error loading translation for locale "${locale}":`, error)
    return {}
  }
}

export const useTranslationServerMiddleware = async (event: H3Event, defaultLocale?: string, currentLocale?: string) => {
  const { getTranslation, loadTranslations, hasGeneralTranslation } = useTranslationHelper()

  const locale = (
    currentLocale
    || event.context.params?.locale
    || getQuery(event)?.locale
    || getCookie(event, 'user-locale')
    || event.headers.get('accept-language')?.split(',')[0]
    || defaultLocale
    || 'en').toString()

  if (!hasGeneralTranslation(locale)) {
    const translations = await fetchTranslations(locale)
    await loadTranslations(locale, translations)
  }

  function t(key: string): string {
    const translation = getTranslation<string>(locale, 'index', key)
    return translation ?? key
  }

  return t
}
