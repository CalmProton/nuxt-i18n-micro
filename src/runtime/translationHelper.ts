import type { Translations } from './plugins/01.plugin'

const generalLocaleCache: { [key: string]: Translations } = {}
const routeLocaleCache: { [key: string]: Translations } = {}
const dynamicTranslationsCaches: { [key: string]: Translations }[] = []

const serverTranslationCache: { [index: string]: Map<string, Translations | unknown> } = { }
const serverTranslationInit: { [index: string]: boolean } = { }

function deepClone<T>(value: T): T {
  if (Array.isArray(value)) {
    return value.slice() as T
  }
  else if (typeof value === 'object' && value !== null) {
    return { ...value } as T
  }
  return value
}

function findTranslation<T = unknown>(translations: Translations | null, parts: string[]): T | null {
  let value: string | number | boolean | Translations | unknown | null = translations

  for (let i = 0; i < parts.length; i++) {
    if (value && typeof value === 'object' && parts[i] in value) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      value = value[parts[i]]
    }
    else {
      return null
    }
  }

  if (typeof value === 'object' && value !== null) {
    return deepClone(value) as T
  }

  return value as T ?? null
}

export function useTranslationHelper() {
  return {
    hasCache(locale: string, page: string) {
      return serverTranslationInit[`${locale}:${page}`] ?? false
    },
    getCache(locale: string, routeName: string) {
      return serverTranslationCache[`${locale}:${routeName}`]
    },
    setCache(locale: string, routeName: string, cache: Map<string, Translations | unknown>) {
      serverTranslationCache[`${locale}:${routeName}`] = cache
      serverTranslationInit[`${locale}:index`] = true
      serverTranslationInit[`${locale}:${routeName}`] = true
    },
    margeTranslation(locale: string, routeName: string, newTranslations: Translations) {
      if (!routeLocaleCache[`${locale}:${routeName}`]) {
        console.error(`marge: route ${locale}:${routeName} not loaded`)
      }
      routeLocaleCache[`${locale}:${routeName}`] = {
        ...routeLocaleCache[`${locale}:${routeName}`],
        ...newTranslations,
      }
    },
    margeGlobalTranslation(locale: string, newTranslations: Translations) {
      if (!generalLocaleCache[`${locale}`]) {
        console.error(`marge: route ${locale} not loaded`)
      }
      generalLocaleCache[locale] = {
        ...generalLocaleCache[locale],
        ...newTranslations,
      }
    },
    hasGeneralTranslation(locale: string) {
      return !!generalLocaleCache[locale]
    },
    hasPageTranslation(locale: string, routeName: string) {
      return !!routeLocaleCache[`${locale}:${routeName}`]
    },
    getTranslation: <T = unknown>(locale: string, routeName: string, key: string, useCache: boolean): T | null => {
      const cacheKey = `${locale}:${routeName}`
      if (useCache && serverTranslationCache[cacheKey]) {
        const cached = serverTranslationCache[cacheKey].get(key)
        if (cached) {
          return cached as T
        }
      }

      const parts = key.split('.')

      let result: null | T = null

      if (dynamicTranslationsCaches.length) {
        for (const dynamicCache of dynamicTranslationsCaches) {
          const value = findTranslation<T>(dynamicCache[locale], parts)
          if (value !== null) {
            result = value
          }
        }
      }

      if (!result) {
        // First, try to find the translation in the route-specific cache
        const value = findTranslation<T>(routeLocaleCache[`${locale}:${routeName}`], parts)
        if (value !== null) {
          result = value
        }
      }

      if (!result) {
        const value = findTranslation<T>(generalLocaleCache[locale], parts)

        if (value !== null) {
          return value
        }
      }

      if (useCache && result) {
        if (!serverTranslationCache[cacheKey]) {
          serverTranslationCache[cacheKey] = new Map<string, Translations>()
        }

        serverTranslationCache[cacheKey].set(key, result)
      }

      return result
    },
    loadPageTranslations: async (locale: string, routeName: string, translations: Translations): Promise<void> => {
      try {
        routeLocaleCache[`${locale}:${routeName}`] = { ...translations }
        serverTranslationInit[`${locale}:${routeName}`] = true
      }
      catch (error) {
        console.error(`Error loading translations for ${locale} and ${routeName}:`, error)
      }
    },
    loadTranslations: async (locale: string, translations: Translations): Promise<void> => {
      try {
        generalLocaleCache[locale] = { ...translations }
        serverTranslationInit[`${locale}:index`] = true
      }
      catch (error) {
        console.error(`Error loading translations for general ${locale}:`, error)
      }
    },
  }
}
