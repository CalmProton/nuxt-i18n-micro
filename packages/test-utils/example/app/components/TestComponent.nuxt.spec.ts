import path from 'node:path'
import fs from 'node:fs'
import { mountSuspended, renderSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, beforeEach } from 'vitest'
import { screen } from '@testing-library/vue'
import { setTranslationsFromJson } from '../../tests/unit-setup'

import TestComponent from './TestComponent.vue'

describe('TestComponent component', () => {
  beforeEach(async () => {
    const fileContent = fs.readFileSync(path.join(__dirname, '../../locales/en-GB.json')).toString()
    await setTranslationsFromJson('en', JSON.parse(fileContent))
  })
  it('renders correctly', async () => {
    const result = await mountSuspended(TestComponent)

    const html = result.html()

    expect(html).toMatchSnapshot()
  })

  it('renders the prop message', async () => {
    const component = await mountSuspended(TestComponent, {
      props: {
        message: 'Test message',
      },
    })

    expect(component.html()).toContain('Test message')
  })

  it('shows the default message if no message prop is passed', async () => {
    await renderSuspended(TestComponent)

    const component = screen.getByTestId('message')

    // This should be "The default message" but we are mocking $t to render the key only
    expect(component.textContent).toContain('The default message')
  })
})
