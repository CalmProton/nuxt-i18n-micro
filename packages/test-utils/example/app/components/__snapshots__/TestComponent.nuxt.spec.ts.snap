// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TestComponent component > renders correctly 1`] = `
"<div>
  <p>Test from component: Welcome translated</p>
  <p data-testid="message"> Test message from props: The default message</p>
</div>"
`;
